In [2]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder

# Sample dataset (list of transactions)
dataset = [
    ['Milk', 'Bread', 'Eggs'],
    ['Bread', 'Eggs', 'Butter'],
    ['Milk', 'Eggs'],
    ['Milk', 'Bread', 'Eggs', 'Butter'],
    ['Milk', 'Bread'],
    ['Bread', 'Eggs'],
    ['Milk', 'Bread', 'Butter'],
]

# Convert the dataset into a one-hot encoded format
te = TransactionEncoder()
te_ary = te.fit(dataset).transform(dataset)
df = pd.DataFrame(te_ary, columns=te.columns_)
print(df)


   Bread  Butter   Eggs   Milk
0   True   False   True   True
1   True    True   True  False
2  False   False   True   True
3   True    True   True   True
4   True   False  False   True
5   True   False   True  False
6   True    True  False   True


## Perform Frequent Itemset Mining using Apriori
Now, we'll use the Apriori algorithm to mine frequent itemsets from the one-hot encoded dataset.

In [3]:
from mlxtend.frequent_patterns import apriori, association_rules

# Define the minimum support threshold (e.g., 0.4 means an itemset must appear in at least 40% of transactions)
min_support = 0.4

# Perform frequent itemset mining using Apriori
frequent_itemsets = apriori(df, min_support=min_support, use_colnames=True)

print("Frequent Itemsets:")
print(frequent_itemsets)

Frequent Itemsets:
    support         itemsets
0  0.857143          (Bread)
1  0.428571         (Butter)
2  0.714286           (Eggs)
3  0.714286           (Milk)
4  0.428571  (Butter, Bread)
5  0.571429    (Eggs, Bread)
6  0.571429    (Milk, Bread)
7  0.428571     (Eggs, Milk)


## Generating Association Rules
we'll use the frequent itemsets to generate association rules and calculate various association metrics such as confidence and lift.

In [8]:
# Generate association rules with minimum confidence threshold (e.g., 0.6)
min_confidence = 0.6
association_rules_df = association_rules(frequent_itemsets, metric="confidence", min_threshold=min_confidence)

print("\nAssociation Rules:")
print(association_rules_df[['antecedents', 'consequents', 'support', 'confidence','lift']])



Association Rules:
  antecedents consequents   support  confidence      lift
0    (Butter)     (Bread)  0.428571    1.000000  1.166667
1      (Eggs)     (Bread)  0.571429    0.800000  0.933333
2     (Bread)      (Eggs)  0.571429    0.666667  0.933333
3      (Milk)     (Bread)  0.571429    0.800000  0.933333
4     (Bread)      (Milk)  0.571429    0.666667  0.933333
5      (Eggs)      (Milk)  0.428571    0.600000  0.840000
6      (Milk)      (Eggs)  0.428571    0.600000  0.840000


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


# Interpretation of the Output : 



The association rules show that buying Butter strongly predicts the purchase of Bread, with a perfect confidence of 1.0 and a positive lift (1.17), indicating a strong relationship. 

Eggs and Milk, when bought, moderately predict the purchase of Bread and each other, but with a lower lift (less than 1), suggesting a weaker or potentially negative association.